File: core/src/main/java/com/triangleleft/flashcards/mvp/vocabular/VocabularyWordPresenter.java
package com.triangleleft.flashcards.mvp.vocabular;

import com.annimon.stream.Optional;
import com.triangleleft.flashcards.mvp.common.di.scope.FragmentScope;
import com.triangleleft.flashcards.mvp.common.presenter.AbstractPresenter;
import com.triangleleft.flashcards.service.vocabular.VocabularyWord;
import com.triangleleft.flashcards.util.FunctionsAreNonnullByDefault;

import javax.inject.Inject;

@FunctionsAreNonnullByDefault
@FragmentScope
public class VocabularyWordPresenter extends AbstractPresenter<IVocabularyWordView> {

    @Inject
    public VocabularyWordPresenter() {
        super(IVocabularyWordView.class);
    }

    public void showWord(Optional<VocabularyWord> word) {
        if (word.isPresent()) {
            applyState(view -> view.showWord(word.get()));
        } else {
            applyState(IVocabularyWordView::showEmpty);
        }
    }
}

File: app/src/main/kotlin/com/triangleleft/flashcards/ui/vocabular/VocabularyWordPresenter.kt
package com.triangleleft.flashcards.ui.vocabular

import com.triangleleft.flashcards.di.scope.FragmentScope
import com.triangleleft.flashcards.service.vocabular.VocabularyWord
import com.triangleleft.flashcards.ui.ViewEvent
import com.triangleleft.flashcards.ui.common.ViewAction
import com.triangleleft.flashcards.ui.common.presenter.AbstractRxPresenter
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject
import javax.inject.Named

@FragmentScope
class VocabularyWordPresenter @Inject constructor(
        @Named(AbstractRxPresenter.UI_SCHEDULER) scheduler: Scheduler
) : AbstractRxPresenter<VocabularyWordView, VocabularyWordViewState, ViewEvent>(scheduler) {

    private val words = PublishSubject.create<VocabularyWord?>()
    lateinit var disposable: CompositeDisposable

    override fun onCreate(savedViewState: VocabularyWordViewState?) {
        super.onCreate(savedViewState)

        val state = savedViewState ?: VocabularyWordViewState(null)

        words.map { word -> ViewAction<VocabularyWordViewState> { state -> state.copy(word = word) } }
                .scan(state) { state, action -> action.reduce(state) }
                .distinctUntilChanged()
                .subscribe {
                    viewStates.onNext(it)
                }
    }

    fun showWord(word: VocabularyWord?) {
        words.onNext(word)
    }
}
